src/pandas_profiling/model/dataframe_wrappers.py

```diff
@@ -0,0 +1,565 @@
+
+    It implements the get_pandas_df method to enable singledispatch wrapped functions to
+    operate on the internal pandas dataframe without breaking abstraction
+
+    """
+
+    def __init__(self, df):
+        super().__init__()
+        # self.df holds the underlying data object
+        self.df = df
+
+    @staticmethod
+    def validate_same_type(obj) -> bool:
+        """
+        Check if pandas dataframe using isinstance. More pythonic way of checking as opposed to spark type check.
+        Possible because its cheap to import pandas dataframe type.
+
+        Args:
+            obj: check if object is a spark dataframe
+
+        Returns: True if the __module__ and __name__ of object matches spark dataframe, else false
+
+        """
+        return isinstance(obj, pd.DataFrame)
+
+    @staticmethod
+    def preprocess(df: pd.DataFrame) -> pd.DataFrame:
+        """Preprocess the dataframe
+
+        - Appends the index to the dataframe when it contains information
+        - Rename the "index" column to "df_index", if exists
+        - Convert the DataFrame's columns to str
+
+        Args:
+            df: the pandas DataFrame
+
+        Returns:
+            The preprocessed DataFrame
+        """
+
+        # Treat index as any other column
+        if (
+            not pd.Index(np.arange(0, len(df))).equals(df.index)
+            or df.index.dtype != np.int64
+        ):
+            df = df.reset_index()
+
+        # Rename reserved column names
+        df = rename_index(df)
+
+        # Ensure that columns are strings
+        df.columns = df.columns.astype("str")
+        return df
+
+    @property
+    def columns(self) -> List[str]:
+        return self.df.columns
+
+    @property
+    def empty(self) -> bool:
+        return self.df.empty
+
+    @property
+    def n_rows(self) -> int:
+        return len(self.df)
+
+    def get_duplicate_rows_count(self, subset) -> int:
+        """
+
+        Args:
+            subset:
+            keep:
+
+        Returns:
+
+        """
+        return len(self.df[self.df.duplicated(subset=subset, keep="first")])
+
+    def dropna(self, subset) -> "PandasDataFrame":
+        return PandasDataFrame(self.df.dropna(subset=subset))
+
+    def groupby_get_n_largest_dups(self, columns, n) -> pd.DataFrame:
+        return (
+            self.df[self.df.duplicated(subset=columns, keep=False)]
+            .groupby(columns)
+            .size()
+            .reset_index(name="count")
+            .nlargest(n, "count")
+        )
+
+    def __len__(self) -> int:
+        return self.n_rows
+
+    def get_memory_usage(self, deep=False) -> pd.Series:
+        return self.df.memory_usage(deep=deep).sum()
+
+    def __getitem__(self, key):
+        return self.df[key]
+
+    def get_pandas_df(self) -> pd.DataFrame:
+        return self.df
+
+    def head(self, n) -> pd.DataFrame:
+        """
+        TODO - should this return a pd.DataFrame or PandasDataFrame?
```

Comment:
pd.DataFrame, there currently is no double usage of the wrapper's result

Reply:
ok!